### PR TITLE
hpac_version_ble

### PR DESCRIPTION
**Pull request by sam412081go**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X ] Code fix
- [ ] Feature
- [ ] Documentation

This patch is for showing HPAC version of BLE device

### CODE REVIEW

File: plugins/pixart-rf/fu-pxi-ble-device.c [hunk 1]
```diff
@@ -841,7 +842,19 @@ fu_pxi_ble_device_fw_get_info(FuPxiBleDevice *self, GError **error)
...
+	} else {
+		if (!fu_memread_uint16_safe(res,
+					    FU_PXI_BLE_DEVICE_OTA_BUF_SZ,
+					    9,
+					    &hpac_ver,
+					    G_BIG_ENDIAN,
+					    error))
+			return FALSE;
+
+		version_str = fu_pxi_hpac_version_info_parse(&hpac_ver);
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-ble-device.c:856]: why parsing as a pointer?

File: plugins/pixart-rf/fu-pxi-common.c [hunk 2]
```diff
@@ -123,3 +123,18 @@ fu_pxi_composite_receiver_cmd(guint8 opcode,
...
+	g_autoptr(GByteArray) hpac_ver_buf = g_byte_array_new();
+	fu_byte_array_append_uint8(hpac_ver_buf, '0' + (*hpac_ver / 10000));
+	fu_byte_array_append_uint8(hpac_ver_buf, '0' + ((*hpac_ver / 1000) % 10));
+	fu_byte_array_append_uint8(hpac_ver_buf, 0x2e);
+	fu_byte_array_append_uint8(hpac_ver_buf, '0' + ((*hpac_ver / 100) % 10));
+	fu_byte_array_append_uint8(hpac_ver_buf, '0' + ((*hpac_ver / 10) % 10));
+	fu_byte_array_append_uint8(hpac_ver_buf, 0x2e);
+	fu_byte_array_append_uint8(hpac_ver_buf, 0x30);
+	fu_byte_array_append_uint8(hpac_ver_buf, '0' + (*hpac_ver % 10));
+	return g_strndup((gchar *)hpac_ver_buf->data, hpac_ver_buf->len);
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-common.c:139]: I think something like `return g_strdup_printf("%u.%u.%u", hpac_ver.....` is much more obvious -- no?

**Comment by sam412081go** [comment]
Hi @hughsie ,

I have encountered a problem.

In the newest commit: 446503b5c1363d42c891c6a53d70bf10fa8ab785 "Fix a fuzzing timeout in the HID descriptor parser" ,
I use "fwupdtool get-devices --plugins=pixart_rf" to get my BLE device info, but fail.

It's shows:
09:41:30.868 FuEngine             failed to add device /sys/devices/virtual/misc/uhid/0005:03F0:114C.0026/hidraw/hidraw5: failed to OTA check retransmit: failed after 30 retries: ioctl error: Input/output error [5]

But in parent commit 81c2a08ef2f2e4ce3eaba3e6babe2b66f12bdbf0 "trivial: Make clearing the local HID state table more efficient", it is ok.

**Comment by hughsie** [comment]
@sam412081go do you get `table invalid, too many duplicate...` messages if you run fwupdtool with -vv? Does the device enumerate properly if you revert that commit? Can you attach the HID descriptor blob to this issue please?

File: plugins/pixart-rf/fu-pxi-common.c (hunk 2 above)
Comment (sam412081go) [reply plugins/pixart-rf/fu-pxi-common.c:139]: Sorry, May I ask how to use "g_strdup_printf" function ?

The range of uint16 hpac_ver is 0~55539, it means 55,53,9. I need to convert to "55,53,9" from number 55539.

I just tried g_strdup_printf("%u,%u,%u",hpac_ver), it shows "Current version:    191,1,49"
But hpac_ver is 191, so it's should be "Current version:    00.19.01"

File: plugins/pixart-rf/fu-pxi-ble-device.c (hunk 1 above)
Comment (sam412081go) [reply plugins/pixart-rf/fu-pxi-ble-device.c]: Fixed.

**Comment by hughsie** [comment]
> Can you attach the HID descriptor blob to this issue please?

Does `sudo usbhid-dump` work?

**Comment by hughsie** [comment]
Or /sys/bus/hid/devices/THE_DEVICE_ID/report_descriptor perhaps?

File: plugins/pixart-rf/fu-pxi-common.c (hunk 2 above)
Comment (hughsie) [reply plugins/pixart-rf/fu-pxi-common.c:139]: Does something like `return g_strdup_printf("%u%u,%u,%u", hpac_ver / 10000, (hpac_ver / 1000) % 10, ((hpac_ver / 100) % 10), (hpac_ver / 10) % 10, hpac_ver % 10);` work?

Comment (sam412081go) [reply plugins/pixart-rf/fu-pxi-common.c]: @hughsie  Thanks for your help! it's working now. And I have committed the newer one to fix it.

**Comment by sam412081go** [comment]
**Does sudo usbhid-dump work?**

sdrf@OA-109048:~/fwupd_fw2_20230809/fwupd/build$ sudo usbhid-dump
001:002:003:DESCRIPTOR         1694566526.565246
 06 01 FF 09 01 A1 01 85 04 19 00 29 7F 15 00 25
 7F 75 08 95 03 B1 01 85 05 95 03 B1 01 85 06 95
 3E B1 01 85 07 95 44 B1 01 85 08 96 12 04 B1 01
 85 09 95 08 B1 01 85 0A 95 03 B1 01 85 0D 95 0A
 B1 01 85 0E 95 05 B1 01 85 0F 95 05 B1 01 85 10
 95 0F B1 01 85 11 95 0C B1 01 85 12 95 3F B1 01
 C0

001:002:002:DESCRIPTOR         1694566526.565694
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 0E 81 00 C0 05 01 09 80 A1 01 85
 02 19 81 29 83 15 00 25 01 75 01 95 03 81 02 95
 05 81 01 C0 05 0C 09 01 A1 01 85 06 19 00 2A FF
 02 15 00 26 FF 02 75 10 95 01 81 00 C0 05 01 09
 02 A1 01 85 04 15 00 26 FF 7F 09 30 09 31 75 10
 95 02 81 02 C0 06 02 FF 09 01 A1 01 85 03 15 00
 26 FF 00 09 2F 75 08 95 07 81 00 C0 05 01 09 0C
 A1 01 85 05 15 00 25 01 09 C6 95 01 75 01 81 06
 75 07 81 03 C0

001:002:001:DESCRIPTOR         1694566526.566126
 06 00 FF 09 01 A1 01 09 02 15 00 26 FF 00 75 08
 95 40 81 02 09 03 15 00 26 FF 00 75 08 95 40 91
 02 C0

001:002:000:DESCRIPTOR         1694566526.566389
 05 01 09 02 A1 01 09 01 A1 00 05 09 15 00 25 01
 19 01 29 05 75 01 95 05 81 02 95 03 81 01 05 01
 16 01 80 26 FF 7F 09 30 09 31 75 10 95 02 81 06
 15 81 25 7F 09 38 75 08 95 01 81 06 05 0C 0A 38
 02 95 01 81 06 C0 C0

**Or /sys/bus/hid/devices/THE_DEVICE_ID/report_descriptor perhaps?**

sdrf@OA-109048:/sys/bus/hid/devices$ ls -lh
total 0
lrwxrwxrwx 1 root root 0  九  13 08:55 0003:1E7D:2D00.0034 -> ../../../devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.3/0003:1E7D:2D00.0034
lrwxrwxrwx 1 root root 0  九  13 08:55 0003:1E7D:2D00.0035 -> ../../../devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.2/0003:1E7D:2D00.0035
lrwxrwxrwx 1 root root 0  九  13 08:55 0003:1E7D:2D00.0036 -> ../../../devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.1/0003:1E7D:2D00.0036
lrwxrwxrwx 1 root root 0  九  13 08:55 0003:1E7D:2D00.0037 -> ../../../devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.0/0003:1E7D:2D00.0037
lrwxrwxrwx 1 root root 0  九  13 08:53 0005:03F0:114C.0033 -> ../../../devices/virtual/misc/uhid/0005:03F0:114C.0033
lrwxrwxrwx 1 root root 0  九  12 15:04 0018:044E:120A.0001 -> ../../../devices/pci0000:00/0000:00:15.1/i2c_designware.1/i2c-2/i2c-DELL08B7:00/0018:044E:120A.0001

My Device is "0005:03F0:114C.0033". It is a Bluetooth mouse device, not usb device.

**Comment by sam412081go** [comment]
**Or /sys/bus/hid/devices/THE_DEVICE_ID/report_descriptor perhaps?**

sdrf@OA-109048:/sys/bus/hid/devices/0005:03F0:114C.003A$ ls
country  driver  hidraw  input  modalias  power  report_descriptor  subsystem  uevent

I use sublime to open "report_descriptor", the content shows below:

0501 0902 a101 8502 0901 a100 0509 1901
2908 1500 2501 9508 7501 8102 0501 1601
8026 ff7f 7510 9502 0930 0931 8106 1581
257f 7508 9501 0938 8106 050c 0a38 0295
0181 06c0 c005 0109 06a1 0185 0105 0719
e029 e715 0025 0175 0195 0881 0295 0575
0105 0819 0129 0591 0295 0175 0391 0195
0a75 0815 0026 ff00 0507 1900 2aff 0081
00c0 050c 0901 a101 8503 1900 2a9c 0215
0026 9c02 7510 9501 8100 c005 0109 80a1
0185 041a 8100 2a83 0015 0025 0175 0195
0381 0295 0581 01c0 0600 ff09 01a1 0125
ff75 0895 1485 0509 01b1 02c0 0605 ff09
01a1 0185 0615 0026 ff00 7508 9514 0902
9102 7508 9514 0903 8102 c006 00ff 0901
a101 1500 26ff 0035 0046 ff00 7508 9514
8507 0901 8102 c006 01ff 0901 a101 7508
9514 8508 0901 b102 c006 02ff 0901 a101
7508 9514 8509 0901 b102 c0

**Comment by hughsie** [comment]
> I use sublime to open "report_descriptor", the content shows below:

Perfect, thanks! Fixed in https://github.com/fwupd/fwupd/commit/71c15c0a66a24ae5129a2d4137153191c7a0e2b4 -- sorry for the trouble.

**Review by hughsie — DISMISSED** [review]

File: plugins/pixart-rf/fu-pxi-firmware.c [hunk 3]
```diff
@@ -130,10 +132,24 @@ fu_pxi_firmware_parse(FuFirmware *firmware,
...
+		if (!fu_memread_uint16_safe(buf,
+					    bufsz,
+					    bufsz - PIXART_RF_FW_HEADER_HPAC_VERSION_POS_FROM_END,
+					    &hpac_ver,
+					    G_BIG_ENDIAN,
+					    error))
+			return FALSE;
+
+		version_raw = hpac_ver;
+		version = fu_pxi_hpac_version_info_parse(hpac_ver);
```
Comment (hughsie) [inline plugins/pixart-rf/fu-pxi-firmware.c:145]: I think you need an #include `#include "fu-pxi-common.h"` at the top of the file for this.

**Review by hughsie — APPROVED** [review]

**Pull request merged**